### PR TITLE
gcc14, gfortran14, gnat14, gccgo14: init at 14.1.0

### PR DESCRIPTION
**Pull request by trofi**
Changes: https://gcc.gnu.org/gcc-14/changes.html

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Review by wegank — APPROVED** [review]

**Comment by wegank** [comment]
Result of `nixpkgs-review pr 309788` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>10 packages failed to build:</summary>
  <ul>
    <li>fastStdenv (gcc14Stdenv)</li>
    <li>gcc14 (gcc_latest)</li>
    <li>gcc14.info (gcc_latest.info)</li>
    <li>gcc14.man (gcc_latest.man)</li>
    <li>gfortran14</li>
    <li>gfortran14.info</li>
    <li>gfortran14.man</li>
    <li>gnat14</li>
    <li>gnat14.info</li>
    <li>gnat14.man</li>
  </ul>
</details>

**Comment by trofi** [comment]
Will have a look at darwin failures later today. Probably stale patches we might have to skip.

**Comment by trofi** [comment]
Result of `nixpkgs-review pr 309788` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>15 packages built:</summary>
  <ul>
    <li>fastStdenv</li>
    <li>gcc14 (gcc_latest)</li>
    <li>gcc14.info (gcc_latest.info)</li>
    <li>gcc14.man (gcc_latest.man)</li>
    <li>gcc14Stdenv</li>
    <li>gccgo14</li>
    <li>gccgo14.info</li>
    <li>gccgo14.man</li>
    <li>gfortran14</li>
    <li>gfortran14.info</li>
    <li>gfortran14.man</li>
    <li>gnat14</li>
    <li>gnat14.info</li>
    <li>gnat14.man</li>
    <li>tests.cc-wrapper.supported</li>
  </ul>
</details>

**Comment by wegank** [comment]
> Will have a look at darwin failures later today. Probably stale patches we might have to skip.

Probably not patches, but a Darwin framework:

```
ld: framework not found CoreServices
clang-16: error: linker command failed with exit code 1 (use -v to see invocation)
make[3]: *** [Makefile:3284: gcov] Error 1
make[3]: *** Waiting for unfinished jobs....
```

**Comment by wegank** [comment]
Result of `nixpkgs-review pr 309788` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>7 packages built:</summary>
  <ul>
    <li>fastStdenv (gcc14Stdenv)</li>
    <li>gcc14 (gcc_latest)</li>
    <li>gcc14.info (gcc_latest.info)</li>
    <li>gcc14.man (gcc_latest.man)</li>
    <li>gfortran14</li>
    <li>gfortran14.info</li>
    <li>gfortran14.man</li>
  </ul>
</details>

**Comment by wegank** [comment]
I think I'll leave x86_64-darwin support as an exercise for ZHF, or if @reckenrode has an idea?

**Comment by reckenrode** [comment]
I’ll try it on my ld64 update branch when I get home. It may be that GCC 14 is broken until the linker can be updated.

I’m pretty sure GNAT should build with it as well. It has a bit of a weird set up with dual linkers.

**Comment by reckenrode** [comment]
> Probably not patches, but a Darwin framework (and yes, again from gettext 0.22)

Is GCC vendoring gettext 0.22? I thought we were still on 0.21.

**Comment by wegank** [comment]
No, they just imported some .m4 files from gettext 0.22 for compatibility.

File: pkgs/development/compilers/gcc/patches/default.nix
Comment (reckenrode) [inline pkgs/development/compilers/gcc/patches/default.nix:1]: Needs definitions of `atLeast14` and `is14`. I actually need that for some of my clang 18 patches.

**Comment by reckenrode** [comment]
I get a different crash with ld 951.9. See the gist for the full crash log from Console.app.

```
ld: warning: object file (/nix/store/krwh63risgpkdw9x3d6ih9jqadl6v3aa-Libsystem-1238.60.2/lib/dylib1.10.5.o) was built for newer macOS version (10.6) than being linked (10.5)
ld: warning: dylib (/nix/store/ali57avprggl9jwd2lsk2x263yjih857-gettext-0.21.1/lib/libintl.dylib) was built for newer macOS version (10.12) than being linked (10.5)
/nix/store/ym3pmv0v68h7qgn5466922djjxa4kisw-clang-wrapper-18.1.5/bin/ld: line 266: 66010 Segmentation fault: 11  /nix/store/5dqbjcafpjmpiwj9bxv89zqiln0fr7qs-cctools-binutils-darwin-1010.6/bin/ld ">
collect2: error: ld returned 139 exit status
```

https://gist.github.com/reckenrode/52a42d70480d5cef9cae8499d49019dd

File: pkgs/development/compilers/gcc/patches/default.nix
Comment (trofi) [reply pkgs/development/compilers/gcc/patches/default.nix:1]: Added unused `is14` / `atLeast14`. Aren't those local variables? Those might be easier to add at the first use.

Comment (reckenrode) [reply pkgs/development/compilers/gcc/patches/default.nix:1]: I can add them in my patches if that’s the preferred approach. I wasn’t sure whether they should be as part of adding a new GCC version.

**Comment by reckenrode** [comment]
I created an ld-wrapper around ld-prime from the latest command-line tools. It also crashes. Next steps are trying to build the bootstrap stage with GCC 13 (in case GCC is somehow causing clang to generate bad code). If that doesn’t work, then try linking with LLD.

```
ld: warning: -single_module is obsolete
ld: warning: duplicate -rpath '/Users/reckenrode/Developer/nixpkgs/outputs/lib/lib' ignored
ld: warning: duplicate -rpath '/nix/store/krwh63risgpkdw9x3d6ih9jqadl6v3aa-Libsystem-1238.60.2/lib' ignored
ld: warning: duplicate -rpath '/Users/reckenrode/Developer/nixpkgs/outputs/lib/lib' ignored
ld: warning: duplicate -rpath '/nix/store/krwh63risgpkdw9x3d6ih9jqadl6v3aa-Libsystem-1238.60.2/lib' ignored
ld: warning: duplicate -rpath '/Users/reckenrode/Developer/nixpkgs/outputs/lib/lib' ignored
ld: warning: duplicate -rpath '/nix/store/krwh63risgpkdw9x3d6ih9jqadl6v3aa-Libsystem-1238.60.2/lib' ignored
ld: warning: support for macOS with 10.5 minimum deployment target is deprecated and will be removed in future release
ld: warning: object file (//nix/store/krwh63risgpkdw9x3d6ih9jqadl6v3aa-Libsystem-1238.60.2/lib/dylib1.10.5.o) was built for newer macOS version (10.6) than being linked (10.5)
0  0x102941807  __assert_rtn + 137
1  0x102952223  ld::passes::stubs::x86_64::classic::LazyPointerAtom::LazyPointerAtom(ld::passes::stubs::Pass&, ld::Atom const&, bool, bool) (.cold.1) + 35
2  0x1028802ef  ld::passes::stubs::x86_64::classic::LazyPointerAtom::LazyPointerAtom(ld::passes::stubs::Pass&, ld::Atom const&, bool, bool) + 431
3  0x10287ca3c  ld::passes::stubs::Pass::makeStub(ld::Atom const&, bool) + 5484
4  0x10287df00  ld::passes::stubs::Pass::process(ld::Internal&) + 1792
5  0x10287e46d  ld::passes::stubs::doPass(Options const&, ld::Internal&) + 109
6  0x1026ce75e  main + 1070
A linker snapshot was created at:
	/tmp/libgcc_s.dylib-2024-05-08-075121.ld-snapshot
ld: Assertion failed: (targetAtom != NULL), function Fixup, file ld.hpp, line 1045.
collect2: error: ld returned 1 exit status
```

**Comment by reckenrode** [comment]
Nope, using gcc13Stdenv also results in a crash.

**Comment by reckenrode** [comment]
> (By the way, `gnat{11,12,13}` also fails on x86_64-darwin on master)

gnat11 is failing because `CC` is set to `clang` instead of `gcc`. It needs the same `stdenv` override in `all-packages.nix` as gnat12 and gnat13. Once that’s set, it fails with the same error as gnat12 and gnat13.

```
configure: error: cannot compute suffix of object files: cannot compile
```

gnat sets `AS_FOR_TARGET` to `${gnat-bootstrap}/bin/gas`, but that path doesn’t exist. The gnat wrapper isn’t picking up `gas` from the special bintools it uses (`bintoolsDualAs`). However,  I wonder if the reasoning for using both assemblers holds anymore. x86_64-darwin now uses the clang assembler by default, so the issue with the GNU assembler in cctools should no longer be a problem.

**Comment by reckenrode** [comment]
Building gnat{11,12,13} on x86_64-darwin is fixed in https://github.com/NixOS/nixpkgs/pull/310450.

**Comment by wegank** [comment]
So let's put x86_64-darwin in `badPlatforms` and merge this, if there's no workaround yet?

**Comment by reckenrode** [comment]
> So let's put x86_64-darwin in `badPlatforms` and merge this, if there's no workaround yet?

I think I found the issue. If I manually the command-line linking `libgcc_s.dylib` to use `-mmacosx-version-min=10.6`, linking succeeds. Looking at `config.host`, GCC is defaulting to building for macOS 10.5 as the minimum instead of 10.8, which works. It’s similar to https://github.com/NixOS/nixpkgs/pull/265935/commits/c6ab64adbfd051383fcd60ac661af7a3aa85b9f1.

The following patch should fix building GCC 14 on Darwin:

```patch
diff --git a/pkgs/development/compilers/gcc/patches/14/libgcc-darwin-detection.patch b/pkgs/development/compilers/gcc/patches/14/libgcc-darwin-detection.patch
new file mode 100644
index 000000000000..8a4ccd1b8d84
--- /dev/null
+++ b/pkgs/development/compilers/gcc/patches/14/libgcc-darwin-detection.patch
@@ -0,0 +1,12 @@
+diff -u a/libgcc/config.host b/libgcc/config.host
+--- a/libgcc/config.host	2023-11-05 11:01:55.778638446 -0500
++++ b/libgcc/config.host	2023-11-05 11:07:29.405103979 -0500
+@@ -227,7 +227,7 @@
+   tmake_file="$tmake_file t-slibgcc-darwin"
+   # newer toolsets produce warnings when building for unsupported versions.
+   case ${host} in
+-    *-*-darwin1[89]* | *-*-darwin2* )
++    *-*-darwin1[89]* | *-*-darwin2* | *-*-darwin)
+       tmake_file="t-darwin-min-8 $tmake_file"
+       ;;
+     *-*-darwin9* | *-*-darwin1[0-7]*)
diff --git a/pkgs/development/compilers/gcc/patches/default.nix b/pkgs/development/compilers/gcc/patches/default.nix
index b170561ce624..a89bc17a1dd8 100644
--- a/pkgs/development/compilers/gcc/patches/default.nix
+++ b/pkgs/development/compilers/gcc/patches/default.nix
@@ -138,6 +140,9 @@ in
 
 ## Darwin
 
+# Fixes detection of Darwin on x86_64-darwin. Otherwise, GCC uses a deployment target of 10.5, which crashes ld64.
+++ optional (atLeast14 && stdenv.isDarwin && stdenv.isx86_64) ../patches/14/libgcc-darwin-detection.patch
+
 # Fix detection of bootstrap compiler Ada support (cctools as) on Nix Darwin
 ++ optional (atLeast12 && stdenv.isDarwin && langAda) ./ada-cctools-as-detection-configure.patch
 
 
```

**Comment by trofi** [comment]
I completely forgot that some of the conditional patches need to be explicitly moved to `gcc14`.

Added as:

```diff
--- a/pkgs/development/compilers/gcc/patches/default.nix
+++ b/pkgs/development/compilers/gcc/patches/default.nix
@@ -65,6 +65,7 @@ in
 ++ optionals (noSysDirs) (
   [(if atLeast12 then ./gcc-12-no-sys-dirs.patch else ./no-sys-dirs.patch)] ++
   ({
+    "14" = [ ./13/no-sys-dirs-riscv.patch ./13/mangle-NIX_STORE-in-__FILE__.patch ];
     "13" = [ ./13/no-sys-dirs-riscv.patch ./13/mangle-NIX_STORE-in-__FILE__.patch ];
     "12" = [ ./no-sys-dirs-riscv.patch ./12/mangle-NIX_STORE-in-__FILE__.patch ];
     "11" = [ ./no-sys-dirs-riscv.patch ];
```

**Comment by reckenrode** [comment]
Does gnat 14 need the same fixes from https://github.com/NixOS/nixpkgs/pull/310450?

**Comment by wegank** [comment]
It's missing `CoreServices` again...

**Pull request merged**